### ISSUE TITLE
[FEAT] Add Regex/partial-string search to more fields

### ISSUE THREAD

**Issue by chaoran-chen** [pathoplexus/pathoplexus#79]
@emmahodcroft wrote in https://github.com/loculus-project/loculus/issues/2576#issue-2486920360:

Currently (I understand) only `authors` has partial-match string search. It would be great to expand this to other fields.

Off the top of my head (but please others comment to recommend more):
- submissionID
- authorAffiliations
- specimenCollectorSampleId

@corneliusroemer wrote in https://github.com/loculus-project/loculus/issues/1742#issue-2275572466:

Substring search will be useful for at least the following metadata types:
- Authors
- Sequence/Isolate names
- Group name
- Author affiliation

**Comment by emmahodcroft** [comment]
Is there a reason not to add this for just about all free-text fields? (Ex not numeric/date fields but all others?)

**Comment by emmahodcroft** [comment]
Accession & genbank accession - talked about at length in 18 Nov Dev meeting --- not possible because we allow a list of values
 - autocomplete possibly a better solution

Emma's Potential list of columns to add substring search to:

- Sequence name (?) - debatable
- Display name (?) - debatable

- Version comment
- Author affiliations
- Authors
- Body product?
- Consensus sequence software name
- Environmental material
- Exposure details
- Exposure event
- Exposure setting
- Food Product
- Host disease
- Host health outcome
- Host health state
- Host vaccination status
- Passage method
- Presampling activity
- Previous infection (disease)
- Previous infection (organism)
- Purpose of sampling (?)
- Purpose of sequencing (?)
- Quality control method name
- Raw sequence data processing method
- Sample type (?)
- Sequenced by - contact name
- Sequenced by
- Sequencing assay type
- Sequencing instrument
- Sequencing protocol
- Signs and symptoms
- Isolate name (?)
- Specimen processing
- Travel history

**Comment by chaoran-chen** [comment]
Hi Emma, thanks for the proposals! I think that we should split the list between partial-string search and autocompletion (which should be as easy to do as turning on the partial-string search), a very spontanous proposal:

**Partial string**

- Sequence name
- Display name
- Version comment
- Author affiliations
- Authors
- Sequenced by - contact name
- Sequenced by
- Isolate name
- Travel history

**Autocompletion**

- Body product
- Consensus sequence software name
- Environmental material
- Food Product
- Host disease
- Host health outcome
- Host health state
- Host vaccination status
- Passage method
- Previous infection (disease)
- Previous infection (organism)
- Purpose of sampling
- Purpose of sequencing
- Quality control method name
- Raw sequence data processing method
- Sample type
- Sequencing assay type
- Sequencing instrument
- Sequencing protocol



**Not sure**

- Exposure details
- Exposure event
- Exposure setting
- Presampling activity
- Signs and symptoms
- Specimen processing

**Comment by emmahodcroft** [comment]
Nice, I didn't realise we could do both! I like the above proposal.

For the not-sure ones, I would guess these are all fairly free-text fields (at least, unlikely to be standardized soon), but maybe partial search would be handy, I'll write out what I could imagine below:

- Exposure details  (maybe searching for "airplane" or "gathering" would be useful, as part of a longer string?)
- Exposure event (Again maybe searching "conference" or "church" as part of some more complex sentence?)
- Exposure setting (this is even less standardized than the above, but maybe partial search still helpful - however, maybe not?)
- Presampling activity (This is drugs or etc that could impact sampling, so again, maybe being able to search "treatment" or "tamiflu")
- Signs and symptoms (maybe searching "cough" "fever" "paralysis" could be helpful?)
- Specimen processing (Maybe searching "pooled" would be helpful?)

**Comment by chaoran-chen** [comment]
@emmahodcroft, just to avoid any potential miscommunication: it's either autocompletion or partial string search: so for the field with autocompletion, it will not be possible to do partial searches.

**Comment by emmahodcroft** [comment]
yes sorry, my comment was confusing. I think I didn't realise "both options are available" for search, but I understand we can't have both in the same field - thanks for checking though as I could have definitely meant otherwise! 😁

**Comment by emmahodcroft** [comment]
Ok but I also think if we can add this to columns that may be useful then why not, they could have entries in the future and there's no reason they couldn't be useful straight away.

I don't want to overcomplicate this issue though. Can you specify which columns you'd change/move/add/remove from Chaoran's suggestion above?

From recounting discussion we already had at a previous meeting, we ruled out any kind of accession partial-string match - from accessing Theo's memory, this was to avoid unexpected matches of `OP123456` against `OP12345` typed in, and things like having 1 billion matches for `01` which could crash the browser. 

I don't think this means it couldn't be added in time, but until we can be more confident we handle this I'd rather just keep this PR as simple as possible to get some extra stuff added to improve search.

**Comment by emmahodcroft** [comment]
This is an issue that's low-bar (is relatively easy to do, and easy to modify in future) and would be really good to add. I'd like to try and push it over the hump.

The only fields above that Cornelius listed as 'Partial String' that differ from what Chaoran wrote earlier are:

INSDC accession
Accession
Specimen processing details

For Accession (of any kind) we had previous discussions (see above) about concerns that partial string could crash the browser if too many hits were generated (ex: 01), and also that not all matches are equally meaningful.
I think particularly given the browser-crash issue (too many hits) we should take this carefully and make make a new issue to investigate this.

I'd be happy to add 'Specimen processing details' to the list of 'Partial string'.

From Chaoran's list above we also had some remaining 'not sure'. I would be happy to include these in 'partial string' for now and we can see how useful it may be - or if it speeds getting this through in some form, happy to leave them out and create a new issue to consider what other fields to add (and any to change.

**Comment by corneliusroemer** [comment]
I think re partial string search there isn't much of a difference.

I didn't list anything as auto-complete because we don't have any data for any of these

**Comment by fhennig** [comment]
@theosanderson I see you moved this, could you maybe update the main issue text to reflect more clearly the outcome of the discussion? I'm not quite sure what needs to be done.

**Comment by theosanderson** [comment]
Uh sorry this must have been a screw up on my phone – my bad. Will move it back when back at a computer it was unintentional. Sorry for the bother

**Comment by fhennig** [comment]
Ah alright no worries!
